### PR TITLE
[Fix #33155] Set through target for new records

### LINKED ISSUE #33155: has_one through not available on unpersisted instances
Opening a new issue since the old one was marked as stale and I didn't get to it in time (doesn't seem like I can re-open it).

https://github.com/rails/rails/issues/31975

@adamlogic confirmed that this still reproduces on current master: https://github.com/rails/rails/issues/31975#issuecomment-397656705

Note: I believe the way that this works is by associating with post_id, so if that's the case it makes sense that this would be the behavior, and I suppose this becomes an enhancement request.

`Comment` knows that it has a `Post` associated with it as evidenced by the passed assertion on L66, so is it possible to leverage that relationship to make the `through` association work?

### Steps to reproduce

Script to reproduce: https://gist.github.com/zachwalton/395412f5b7e97837e6273981dbb7faf3

### Expected behavior

`owner` should be not be nil when accessed through the `has_one through` association

### Actual behavior

`owner` is nil when accessed through the `has_one through` association

### System configuration
**Rails version**: 5.1.2

**Ruby version**: 2.4.2

### PR DESCRIPTION
**Pull request by joshuay03**
### Motivation / Background

Fixes #33155

### Detail

Updates `#load_target` in `ActiveRecord::Associations::CollectionAssociation` and `ActiveRecord::Associations::Association` (which is inherited by `ActiveRecord::Associations::SingularAssociation`) to set the through association target using the middle record when the current record has not yet been persisted.

### Additional information

There also seems to have been [another attempted patch](https://github.com/rails/rails/pull/38562) which didn't end up getting merged.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/associations/has_many_through_associations_test.rb [hunk 1]
```diff
@@ -53,6 +53,20 @@ def setup
...
+    subscription_1 = Subscription.new(subscriber: subscriber_1)
+    subscription_2 = Subscription.new(subscriber: subscriber_2)
+    book = Book.new
+    book.subscriptions = [subscription_1, subscription_2]
+
+    assert_predicate subscriber_1, :persisted?
+    assert_predicate subscriber_2, :persisted?
+    assert_predicate book, :new_record?
+    book.subscriptions.each { |subscription| assert_predicate subscription, :new_record? }
+    assert_equal book.subscribers.sort, [subscriber_1, subscriber_2].sort
```
Comment (joshuay03) [inline activerecord/test/cases/associations/has_many_through_associations_test.rb:67]: Before patch:
Screenshot 2024-03-20 at 7 02 21 PM

File: activerecord/test/cases/associations/has_one_through_associations_test.rb [hunk 2]
```diff
@@ -43,6 +43,18 @@ def test_has_one_through_executes_limited_query
...
+  def test_setting_association_on_new_record_sets_through_record
+    club = Club.create!
+    membership = CurrentMembership.new(club: club)
+    member = Member.new
+    member.current_membership = membership
+
+    assert_predicate club, :persisted?
+    assert_predicate member, :new_record?
+    assert_predicate member.current_membership, :new_record?
+    assert_equal club, member.club
```
Comment (joshuay03) [inline activerecord/test/cases/associations/has_one_through_associations_test.rb:55]: Before patch:
Screenshot 2024-03-20 at 7 21 05 PM

**Comment by sromano** [comment]
@joshuay03 and @byroot Thanks for fixing this. I found a bug with nested has_one :through associations in this implementation that can easily fix. I can push a PR tomorrow

```
class ProgramEnrollment < ApplicationRecord
  has_one :person, through: :customer
  has_one :organization, through: :person
end
```

`program_enrollment.organization` will still return `nil` in the same context described in the original bug.

It gets fixed if 
`through_association.target.association(reflection.source_reflection_name).target`
is changed to:
`through_association.target.association(reflection.source_reflection_name).load_target`

**Comment by joshuay03** [comment]
> @joshuay03 and @byroot Thanks for fixing this. I found a bug with nested has_one :through associations in this implementation that can easily fix. I can push a PR tomorrow
> 
> ```
> class ProgramEnrollment < ApplicationRecord
>   has_one :person, through: :customer
>   has_one :organization, through: :person
> end
> ```
> 
> `program_enrollment.organization` will still return `nil` in the same context described in the original bug.

Oh, I did not consider this case. I think I managed to reproduce this but in my case it actually raises an error cause the `source:` needs to be specified on the nested through association, and we incorrectly try to grab the target off the first through association.

> It gets fixed if `through_association.target.association(reflection.source_reflection_name).target` is changed to: `through_association.target.association(reflection.source_reflection_name).load_target`

I don't think we want to load the target here, this fix is only for when the through target has been loaded i.e. is in memory, and the owner is a new record. I think I've got an alternative fix for this, I'll open a PR soon and I would be grateful if you could test your case against that branch.

**Comment by joshuay03** [comment]
https://github.com/rails/rails/pull/53869

**Pull request merged**